### ISSUE TITLE
Idea: a deterministic "ready" signal for Tye applications

### ISSUE THREAD

**Issue by philliphoff** [dotnet/tye#798]
As Tye commands (e.g. `tye run`) are long-running tasks and incur a variable amount of initialization time, it would be useful for automation purposes to be able to understand when those commands have "completed" (i.e. the application is running).  Also, such a signal is best done in an asynchronous manner rather than require polling by the automation (which would then incur latency).  There are a number of ways this can be accomplished (with varying amount of effort):

 - **Shortest term:** write a definitive "ready" string to standard output. This is similar to the approach used by automation for  ASP.NET applications which typically wait for a "listening on port..." string.  The downside to this approach is that it can be complicated by localization efforts, terminal formatting, and output buffering.
 - **Medium term:** publish the status of a Tye application via the Tye dashboard API.  The downside to this approach is (1) knowing where and when the API endpoint is ready and (2) knowing when that status changes.
 - **Longer-term:** provide an asynchronous mechanism to both identify running Tye applications as well as their state changes.  This could be as fancy as using mDNS to advertise Tye applications on the network and then SignalR-style callbacks in the API endpoint for state changes.

**Comment by areller** [comment]
As of now, you can define a readiness probe for you service (see https://github.com/dotnet/tye/blob/master/docs/reference/schema.md#liveness-and-readiness) and then use the HTTP API that tye exposes and poll the `State` of the replicas of your service.

I don't think that there is a documentation for the API, but you can look at here https://github.com/dotnet/tye/blob/master/src/Microsoft.Tye.Hosting/TyeDashboardApi.cs to see which methods are exposed.

**Comment by philliphoff** [comment]
In my case, I'm less interested in knowing the readiness of each individual service so much as the completion of the *orchestration* performed by Tye.  That is, my automation purposes simply want to know when Tye thinks it's done initializing all of the extension services, sidecars, and services.

Looking at the readiness of all services *could* work, but likely assumes knowledge of the application services (such as that they support readiness probes in the first place) and may not work in general.  Even then, it would still be helpful to have a non-polling mechanism for notifications of state changes (to reduce the latency of the automation).

**Comment by goblinfactory** [comment]
Maybe ready can be something you subscribe to? I.e. I may only be interested when A,D, H and J are all ready, but don't care about B, C, E, G and I? Foo.WhenAll.Ready(A, D, H, J)

**Comment by areller** [comment]
> In my case, I'm less interested in knowing the readiness of each individual service so much as the completion of the _orchestration_ performed by Tye. That is, my automation purposes simply want to know when Tye thinks it's done initializing all of the extension services, sidecars, and services.
> 
> Looking at the readiness of all services _could_ work, but likely assumes knowledge of the application services (such as that they support readiness probes in the first place) and may not work in general. Even then, it would still be helpful to have a non-polling mechanism for notifications of state changes (to reduce the latency of the automation).

I can agree with having a non polling mechanism. I don't know if I like the idea of sending signals via the CLI stdout, the text there is always subject to change, but maybe something that's based on websockets could work.

I still think that you should be using readiness probes to determine when a service is ready. Even if all the sidecar containers are up and all the ports are open, the service could still be unprepared to accept requests. If your services are straightforward and what I said doesn't apply to them, you could create a very basic readiness probe that points at `/`

**Comment by philliphoff** [comment]
> I can agree with having a non polling mechanism. I don't know if I like the idea of sending signals via the CLI stdout, the text there is always subject to change, but maybe something that's based on websockets could work.

Agree, `stdout` signals would only be a short-term solution due to their fragility (due to versioning as well as localization).

> I still think that you should be using readiness probes to determine when a service is ready. Even if all the sidecar containers are up and all the ports are open, the service could still be unprepared to accept requests. If your services are straightforward and what I said doesn't apply to them, you could create a very basic readiness probe that points at /

For my purposes, I'm less interested in whether the services themselves are ready to accept traffic and more whether Tye considers itself done with its orchestration.  The former can be used as an approximation for the latter, but also makes assumptions about Tye services that I don't necessarily want to make in general (such as that they've been authored with any sort of readiness probes).

**Comment by goblinfactory** [comment]
A simple pattern that bypasses all these issues could be where Tye starts a queued bus of some kind, then starts the services only after the bus is ready to accept messages, then it doesn't matter if services are "ready" or not. They'll process messages in the queue when they start.
